### ISSUE TITLE
Tensorflow on android: directly build app in python?

### ISSUE THREAD

**Issue by michaelhuang74** [tensorflow/tensorflow#5458]
The sample app given by google for tensorflow on android is written in C++.

I have a tensorflow application written in python. This application currently runs on desktop. I want to move the application to android platform. Can I use bazel to build the application that is written in python directly for android? Thanks.

**Comment by andrewharp** [comment]
@michaelhuang74 No, the short answer is that it is not possible (or at least supported) to run any TensorFlow python code on Android. However, you may of course reuse your trained GraphDef proto (after running [optimize_for_inference.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py) on it to bake in the checkpoint weights), and if your interaction with TensorFlow is not too complex you should be able to fairly easily hook up your inputs and outputs to achieve the same effect in either Java or C++.

Can you clarify a bit more exactly what you're trying to achieve? Do you have an existing app you wanted to hook the python code into, or did you want to run it as a standalone program? In the former case, you'll need to convert it to use TensorFlowInferenceInterface.java. In the latter, you'll want to convert to C++ as seen in tensorflow/tools/benchmark:benchmark_model.

**Comment by michaelhuang74** [comment]
**@andrewharp**  Thank you very much for your reply.

The python application that I want to move to android is evaluate.py at https://github.com/lengstrom/fast-style-transfer. 

evaluate.py will take a model file (checkpoint file) generated by the style.py and stylize an input image.  Style.py will stay on desktop computer. I intend to move the evaluate.py to android.

I am new to both tensorflow and application development on android. I appreciate if you can provide some specifics how to modify the sample tensorflow on android demo to port the evaluate.py to android platform. Thanks.

**Comment by michaelhuang74** [comment]
While convert python code to C++ code, we find that the python code uses several python APIs such as tf.nn.tanh, tf.nn.conv2d, tf.nn.relu, tf.nn.moment, tf.nn.conv2d_transpose.  

Where in the tensorflow repository can I find the C++ implementation for the above python APIs? Thanks.

**Comment by andrewharp** [comment]
TF should find these operations automatically when loading your trained GraphDef, which you should use freeze_graph.py to create in conjunction with your checkpoint file. You do not need to refer to them by name in the Android app, you just need to set the input and output nodes similar to how the demo does it in [ClassifierActivity](https://github.com/tensorflow/tensorflow/blob/a2eba827afa870f7db0e0a582c630125a622e725/tensorflow/examples/android/src/org/tensorflow/demo/ClassifierActivity.java#L51)

Apologies for the delayed response! This is not a TF issue though -- you may have better luck on with these sorts of queries on stackoverlow.

**Comment by dust0x** [comment]
@andrewharp I have written a slightly more complex architecture in TF which is ConvNet + LSTMs with _visual attention_. Closer to [Im2Text](https://github.com/tensorflow/models/tree/master/im2txt) than your classifier models. Say I checkpoint my graph and rewrite ClassifierActivity as instructed in your previous comment, will I be able to make it work on an Android device? 

If not, what other changes might I have to make?

**Comment by PranjalSahu** [comment]
Is is possible to execute tf.gradients in android environment.
I have to take gradients in Android application.
What is the best way to do so.

Thanks
